src/nvim/eval/encode.c

```diff
@@ -0,0 +1,1258 @@
+static const void *last_p_enc = NULL;
+
+/// Conversion setup for converting from last_p_enc to UTF-8
+static vimconv_T p_enc_conv = {
+  .vc_type = CONV_NONE,
+};
+
+/// Escape sequences used in JSON
+static const char escapes[][3] = {
+  [BS] = "\\b",
+  [TAB] = "\\t",
+  [NL] = "\\n",
+  [CAR] = "\\r",
+  ['"'] = "\\\"",
+  ['\\'] = "\\\\",
+};
+
+static const char xdigits[] = "0123456789ABCDEF";
+
+/// Convert given string to JSON string
+///
+/// @param[out]  gap  Garray where result will be saved.
+/// @param[in]  buf  Converted string.
+/// @param[in]  len  Converted string length.
+///
+/// @return OK in case of success, FAIL otherwise.
+static inline int convert_to_json_string(garray_T *const gap,
+                                         const char *const buf,
+                                         const size_t len)
+  FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_ALWAYS_INLINE
+{
+  const char *utf_buf = buf;
+  if (utf_buf == NULL) {
+    ga_concat(gap, "\"\"");
+  } else {
+    size_t utf_len = len;
+    char *tofree = NULL;
+    if (last_p_enc != (const void *) p_enc) {
+      p_enc_conv.vc_type = CONV_NONE;
+      convert_setup(&p_enc_conv, p_enc, "utf-8");
+      p_enc_conv.vc_fail = true;
+      last_p_enc = p_enc;
+    }
+    if (p_enc_conv.vc_type != CONV_NONE) {
+      tofree = string_convert(&p_enc_conv, buf, &utf_len);
+      if (tofree == NULL) {
+        EMSG2(_("E474: Failed to convert string \"%s\" to UTF-8"), utf_buf);
+        return FAIL;
+      }
+      utf_buf = tofree;
+    }
+    size_t str_len = 0;
+    // Encode character as \u0000 if
+    // 1. It is an ASCII control character (0x0 .. 0x1F).
+    // 2. &encoding is not UTF-8 and code point is above 0x7F.
+    // 3. &encoding is UTF-8 and code point is printable according to
+    //    utf_printable().
+    // This is done to make it possible to :echo values when &encoding is not
+    // UTF-8.
+#define ENCODE_RAW(p_enc_conv, ch) \
+    (ch >= 0x20 && (p_enc_conv.vc_type == CONV_NONE \
+                    ? utf_printable(ch) \
+                    : ch < 0x7F))
+    for (size_t i = 0; i < utf_len;) {
+      const int ch = utf_ptr2char(utf_buf + i);
+      const size_t shift = (ch == 0? 1: utf_ptr2len(utf_buf + i));
+      assert(shift > 0);
+      i += shift;
+      switch (ch) {
+        case BS:
+        case TAB:
+        case NL:
+        case FF:
+        case CAR:
+        case '"':
+        case '\\': {
+          str_len += 2;
+          break;
+        }
+        default: {
+          if (ch > 0x7F && shift == 1) {
+            EMSG2(_("E474: String \"%s\" contains byte that does not start any "
+                    "UTF-8 character"), utf_buf);
+            return FAIL;
+          } else if ((0xD800 <= ch && ch <= 0xDB7F)
+                     || (0xDC00 <= ch && ch <= 0xDFFF)) {
+            EMSG2(_("E474: UTF-8 string contains code point which belongs "
+                    "to surrogate pairs: %s"), utf_buf + i);
+            return FAIL;
```

Comment:
> Missing xfree(tofree);?

Not actually: these conditions should not happen after string_convert with .vc_fail=true. Though I will add them just in case.